### PR TITLE
bpo-45107: Make LOAD_METHOD_CLASS safer and faster, clean up comments

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
https://bugs.python.org/issue45107

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
I'll ping Mark when this is ready and he's back.

IIUC, `LOAD_METHOD_CLASS` doesn't need to check `cls.__dict__` at all. A single `tp_version_tag` check is sufficient. So we can save on dict calculation and a branch.

File: Python/ceval.c [hunk 1]
```diff
@@ -4517,13 +4518,8 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             PyObject *cls = TOP();
             PyTypeObject *cls_type = Py_TYPE(cls);
             assert(cls_type->tp_dictoffset > 0);
```
Comment (erlend-aasland) [inline Python/ceval.c]: You can now wrap these two lines in `#ifdef Py_DEBUG` to get rid of the unused variable warning :)

**Comment by Fidget-Spinner** [comment]
@erlend-aasland the more I look at this the more iffy my code feels.

I'm trying to think if this could cause any segfaults. E.g. it accesses a non-type object at `offsetoff(PyTypeObject, tp_version_tag`, but the object size is smaller than that location, and it just crashes. I will try to produce some code to test my theory.

Previously the dict check also served as an object identity check since `dk_version` is unique. With that, accessing `tp_version_tag` is always safe.

This is guaranteed to always work though if all objects are bigger than `offsetoff(PyTypeObject, tp_version_tag)`. I dont know if thats the case though. Do you remember?

**Comment by erlend-aasland** [comment]
> I dont know if thats the case though. Do you remember?

I do not, and unfortunately I do not have any time to investigate now :(

**Comment by Fidget-Spinner** [comment]
Thanks for the review Eerlend, ultimately I decided to roll back the changes and make the checks stricter. My tests indicate we lost ``0.1%`` of specialization hits, but the added safety is worth it. I can't prove that accessing `tp_version_tag` will never segfault in all cases.

File: Python/ceval.c [hunk 2]
```diff
@@ -4518,8 +4519,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
             PyObject *dict = *(PyObject **) ((char *)cls + cls_type->tp_dictoffset);
-            // Don't care if no dict -- tp_version_tag should catch anything wrong.
-            DEOPT_IF(dict != NULL && ((PyDictObject *)dict)->ma_keys->dk_version !=
+            // All types should have a dict. dk_version is also an identity check
+            // to ensure `cls` is really a type object. Otherwise, reading
+            // tp_version_tag may segfault.
```
Comment (markshannon) [inline Python/ceval.c]: Reading `tp_version_tag` cannot segfault if `cls` is a `PyTypeObject`.
However, that isn't checked. Not sure how we missed that.
Can you add `DEOPT_IF(!PyType_Check(cls), LOAD_METHOD);` here.

File: Python/ceval.c [hunk 3]
```diff
@@ -4518,8 +4519,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+            DEOPT_IF(dict == NULL || ((PyDictObject *)dict)->ma_keys->dk_version !=
```
Comment (markshannon) [inline Python/ceval.c]: There is no need to check the version of the keys, the type version is sufficient.

**Review by markshannon — COMMENTED** [review]
It looks like we missed a check that the operand of `LOAD_METHOD_CLASS` is actually a class.

File: Python/ceval.c (hunk 3 above)
Comment (Fidget-Spinner) [reply Python/ceval.c]: Perhaps I was trying too hard, but I had the impression that _PyType_Check was slow, so I'd used the `dk_version` to validate.

Now that I've taken a look at _PyType_Check's code, I realized I was naive to think that way :). It's just a ``type->tp_flags & Py_TPFLAGS_TYPE_SUBCLASS`` once inlined. Plenty faster than a dict calculation.

**Comment by markshannon** [comment]
Looks good, thanks.
Don't know what is up with the Windows builds :confused:

**Comment by Fidget-Spinner** [comment]
FWIW, I can't repro the windows build errors locally. It seems to be affecting every PR right now.

**Comment by Fidget-Spinner** [comment]
Thanks Mark and Erlend for the reviews!

**Pull request merged**